Feature/cpu replay buffer

weepingwillowben opened the pull request:
This gives the replay buffers a `store_device` parameter which allows it to store the replay buffer on a different device than it expects input and output to be. The main use case is if you want the replay buffer to be stored in CPU memory while the network and training lives in GPU memory. I measured the performance hit of doing this to be about 10-15% on DQN. 

Right now there is no parameter for `store_device` on the presets, though perhaps there should be.

cpnota commented on all/bodies/vision.py:
```diff
@@ -9,23 +9,55 @@ def __init__(self, agent, size=4, lazy=False):
+            self.from_device = value.device
+        if self.to_device is None:
+            self.to_device = device
+        if self.from_device != value.device or self.to_device != device:
```
I find the usage a little confusing... `device` only has one valid value if `to_device` is set, but can be set to anything if `to_device` is none. Is there any reason not to choose one usage or the other?

weepingwillowben replied:
Ok, I greatly simplified this logic.

cpnota commented on all/core/state.py:
```diff
@@ -291,7 +304,7 @@ def as_output(self, tensor):
 
     def apply_mask(self, tensor):
-        return tensor * self.mask.unsqueeze(-1)
+        return tensor * self.mask.unsqueeze(-1)  # pylint: disable=no-member
```
Don't need the pylint disable because we change the linter

weepingwillowben replied:
Removed

cpnota commented on all/memory/replay_buffer.py:
```diff
@@ -23,14 +23,17 @@ def update_priorities(self, indexes, td_errors):
 # https://github.com/Shmuma/ptan/blob/master/ptan/experience.py
 class ExperienceReplayBuffer(ReplayBuffer):
-    def __init__(self, size, device=torch.device('cpu')):
+    def __init__(self, size, device='cpu', store_device='cpu'):
```
Perhaps `store_device` could default to `None`, in which case `store_device` would just be set to `device`?

weepingwillowben replied:
Changed

cpnota commented on all/bodies/vision.py:
```diff
@@ -9,23 +9,55 @@ def __init__(self, agent, size=4, lazy=False):
         return state.update('observation', torch.cat(self._frames, dim=0))
 
 
+class ToCache:
```
Perhaps it would be good to add a docstring explaining the purpose of this? It probably would not be immediately obvious to somebody reading the code for the first time.

weepingwillowben replied:
Added a docstring

cpnota reviewed (changes requested):
Looks really good overall, this should be very helpful on many machines! The design seems correct. I left a few comments in line, and also have a few high-level comments:

1. Could you extend the unit tests to test some of this functionality? In particular, `state_test.py` and `replay_buffer_test.py`. `FrameStack` also really should have unit tests, but I won't hold you to that as they were already missing.
2. It might be useful to add docstrings in certain places in order to explain the usage.
3. As you commented, for this to be useful, the presets need to be modified. My thought is that a hyperparameter, something like `cpu_replay_buffer=False`, could be added. One of the recent PRs added the ability to set hyperparameters from the command line.

weepingwillowben commented:
1. Added unit tests to state_test.py and replay_buffer_test.py
2. Added docstrings to the cache (and changed the name)
3. Haven't changed the presets yet.

weepingwillowben commented:
Thanks
